### PR TITLE
Feature/shadows

### PR DESCRIPTION
**Pull request by tonifj**
Features:

* Normalized Blinn Phong Lighting corrections
* Shadows from directional light source

### CODE REVIEW

File: Engine/Component/ComponentLight.cpp [hunk 1]
```diff
@@ -11,6 +12,15 @@ ComponentLight::ComponentLight() : Component(nullptr, ComponentType::LIGHT)
...
 
+}
+void ComponentLight::Init()
+{
+
+}
+
+void ComponentLight::Update()
+{
+
```
Comment (Dartemiss) [inline Engine/Component/ComponentLight.cpp]: Why do you define these functions if they are empty?

File: Engine/Component/ComponentLight.h [hunk 2]
```diff
@@ -16,6 +16,22 @@ class ComponentLight : public Component
 		DIRECTIONAL_LIGHT
 	};
 
+	class DirectionalLight {
+	public:
+		Frustum directional_light_camera; //From we'll record the depth buffer
+		GLuint rbo = 0;		//Crear framebuffers necessaris per sa direc_light
```
Comment (Dartemiss) [inline Engine/Component/ComponentLight.h]: All comments should be in english!

File: Engine/Component/ComponentMeshRenderer.cpp [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include "Module/ModuleResourceManager.h"
 #include "Module/ModuleScene.h"
 #include "Module/ModuleTexture.h"
+#include "Module/ModuleCamera.h"
```
Comment (Dartemiss) [inline Engine/Component/ComponentMeshRenderer.cpp]: Code style, includes should be in alphabetical order!

File: Engine/Component/ComponentMeshRenderer.cpp [hunk 4]
```diff
@@ -145,7 +146,7 @@ void ComponentMeshRenderer::AddSpecularUniforms(unsigned int shader_program) con
 	glUniform1i(glGetUniformLocation(shader_program, "material.specular_map"), 2);
 	glUniform4fv(glGetUniformLocation(shader_program, "material.specular_color"), 1, (float*)material_to_render->specular_color);
 	glUniform1f(glGetUniformLocation(shader_program, "material.k_specular"), material_to_render->k_specular);
-	glUniform1f(glGetUniformLocation(shader_program, "material.shininess"), material_to_render->specular_color[3]);
+	//glUniform1f(glGetUniformLocation(shader_program, "material.shininess"), material_to_render->specular_color[3]);
 
 	//Material BRDF variables
 	//glUniform1f(glGetUniformLocation(shader_program, "material.roughness"), material_to_render->roughness);
```
Comment (Dartemiss) [inline Engine/Component/ComponentMeshRenderer.cpp]: Why is this commented? It will be used?

File: Engine/EditorUI/Panel/PanelInspector.cpp [hunk 5]
```diff
@@ -3,6 +3,9 @@
 #include "Main/Application.h"
 #include "Main/GameObject.h"
 #include "Module/ModuleEditor.h"
+#include "Module/ModuleCamera.h"
+#include "Module/ModuleLight.h"
```
Comment (Dartemiss) [inline Engine/EditorUI/Panel/PanelInspector.cpp]: Code style

File: Engine/Module/ModuleCamera.h [hunk 6]
```diff
@@ -40,6 +40,8 @@ class ModuleCamera : public Module
 public:
 	ComponentCamera *scene_camera = nullptr;
 	ComponentCamera* main_camera = nullptr;
+	ComponentCamera* directional_light_camera = nullptr;
+	GameObject *dir_light_game_object = nullptr;
```
Comment (Dartemiss) [inline Engine/Module/ModuleCamera.h]: Code style, GameObject*.

File: Engine/Module/ModuleLight.cpp [hunk 7]
```diff
@@ -2,6 +2,8 @@
 #include "Main/Application.h"
 #include "Module/ModuleProgram.h"
 #include "Main/GameObject.h"
+#include "Module/ModuleRender.h"
+#include "Module/ModuleCamera.h"
```
Comment (Dartemiss) [inline Engine/Module/ModuleLight.cpp]: Code style

**Review by Dartemiss — CHANGES_REQUESTED** [review]
Good job! There is no scene to test the shadows? I've tried to do it manually but i wasn't able to test it.

File: Engine/Component/ComponentLight.cpp (hunk 1 above)
Comment (tonifj) [reply Engine/Component/ComponentLight.cpp]: That is some leftover code, my mistake

File: Engine/Component/ComponentMeshRenderer.cpp (hunk 4 above)
Comment (tonifj) [reply Engine/Component/ComponentMeshRenderer.cpp]: That's some very old leftover code.

**Comment by tonifj** [comment]
I just pushed a scene where you can see a model project a shadow to another

**Comment by Dartemiss** [comment]
imagen
This is what i see on the scene you uploaded

**Comment by Dartemiss** [comment]
https://gyazo.com/17c6f882159b5f1e49130ed905816a61
I can't see the shadows :(

**Comment by tonifj** [comment]
There was a bug where models without specular texture wouldn't render its material correctly, by now, it should be fixed

File: Engine/Component/ComponentCamera.cpp [hunk 8]
```diff
@@ -119,6 +133,11 @@ void ComponentCamera::Delete()
 	App->cameras->RemoveComponentCamera(this);
 }
 
+void ComponentCamera::SetProj(const float4x4& proj_matrix)
+{
+	proj = proj_matrix;
+}
+
```
Comment (mariofv) [inline Engine/Component/ComponentCamera.cpp]: This method is not used in the code, and also changing the projection matrix implies changin more stuf (frustum type and parameters). I feel that this function should not exist or at least it has to be expanded

File: Engine/Component/ComponentCamera.h [hunk 9]
```diff
@@ -138,20 +147,29 @@ class ComponentCamera : public Component
...
+
 private:
-	
-	
-	
 	GLuint msfbo = 0;
 	GLuint msfb_color = 0;
-	GLuint last_recorded_frame_texture = 0;
+
+	GLuint depthfbo = 0;
```
Comment (mariofv) [inline Engine/Component/ComponentCamera.h]: Code style, it should be `depth_fbo`

File: Engine/Component/ComponentLight.cpp [hunk 10]
```diff
@@ -3,6 +3,7 @@
 #include "Main/GameObject.h"
 #include "Module/ModuleLight.h"
 #include "Module/ModuleProgram.h"
+#include "Module/ModuleCamera.h"
```
Comment (marcorod94) [inline Engine/Component/ComponentLight.cpp]: Code Style alphabetical order

File: Engine/Component/ComponentMeshRenderer.cpp [hunk 11]
```diff
@@ -65,14 +64,28 @@ void ComponentMeshRenderer::Render()
...
+
+	if (material_to_render->GetMaterialTexture(Material::MaterialTextureType::SPECULAR) != nullptr)
+		material_to_render->use_specular_map = true;
+	else
+		material_to_render->use_specular_map = false;
+
+	if (material_to_render->GetMaterialTexture(Material::MaterialTextureType::NORMAL) != nullptr)
+		material_to_render->use_normal_map = true;
+	else
+		material_to_render->use_normal_map = false;
```
Comment (mariofv) [inline Engine/Component/ComponentMeshRenderer.cpp]: This can be simplified to
```
material_to_render->use_specular_map = material_to_render->GetMaterialTexture(Material::MaterialTextureType::SPECULAR) != nullptr;
material_to_render->use_normal_map = material_to_render->GetMaterialTexture(Material::MaterialTextureType::NORMAL) != nullptr
```

File: Engine/Component/ComponentMeshRenderer.cpp [hunk 12]
```diff
@@ -112,6 +125,7 @@ void ComponentMeshRenderer::AddDiffuseUniforms(unsigned int shader_program) cons
 	glUniform1i(glGetUniformLocation(shader_program, "material.diffuse_map"), 0);
 	glUniform4fv(glGetUniformLocation(shader_program, "material.diffuse_color"), 1, (float*)material_to_render->diffuse_color);
 	glUniform1f(glGetUniformLocation(shader_program, "material.k_diffuse"), material_to_render->k_diffuse);
+	//glActiveTexture(0);
```
Comment (mariofv) [inline Engine/Component/ComponentMeshRenderer.cpp]: If this commented code is not usefull we should delete it to keep the code clean

File: Engine/Component/ComponentLight.h [hunk 13]
```diff
@@ -16,6 +16,22 @@ class ComponentLight : public Component
...
+		GLuint rbo = 0;					  //Direc_light buffers
+		GLuint fbo = 0;
+		GLuint depth_rbo = 0;
+		GLuint depth_map = 0;
+		GLuint last_recorded_frame_texture = 0;
+
+		float4x4 proj;
+		float4x4 view;
+
+		int SHADOW_WIDTH = 1024;
```
Comment (marcorod94) [inline Engine/Component/ComponentLight.h]: Where do you use this variables

File: Engine/EditorUI/Panel/InspectorSubpanel/PanelComponent.cpp [hunk 14]
```diff
@@ -490,6 +490,12 @@ void PanelComponent::ShowComponentCameraWindow(ComponentCamera *camera)
 		//UndoRedo
 		CheckClickedCamera(camera);
 
+		if (ImGui::Button("test"))
```
Comment (marcorod94) [inline Engine/EditorUI/Panel/InspectorSubpanel/PanelComponent.cpp]: Remove this now that is working

File: Engine/EditorUI/Panel/InspectorSubpanel/PanelComponent.cpp [hunk 15]
```diff
@@ -490,6 +490,12 @@ void PanelComponent::ShowComponentCameraWindow(ComponentCamera *camera)
...
+		{
+
+			APP_LOG_INFO("%s", camera->camera_frustum.ToString().c_str());
+		}
+
```
Comment (mariofv) [inline Engine/EditorUI/Panel/InspectorSubpanel/PanelComponent.cpp]: If this code is not used it should be deleted

File: Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp [hunk 16]
```diff
@@ -270,6 +275,18 @@ bool PanelMaterial::ShowMaterialTextureMap(std::shared_ptr<Material> material, M
...
+			modified_by_user = true;
+		}
+		/* if (ImGui::SliderFloat("Roughness", &material->roughness, 0.f, 1.f))
+		{
+			modified_by_user = true;
+		}
+		if (ImGui::SliderFloat("Metalness", &material->metalness, 0.f, 10.f))
+		{
+			modified_by_user = true;
+		}*/
```
Comment (mariofv) [inline Engine/EditorUI/Panel/InspectorSubpanel/PanelMaterial.cpp]: Same as before, commented code that is not used should be deleted

File: Engine/EditorUI/Panel/PanelConfiguration.cpp [hunk 17]
```diff
@@ -291,6 +294,26 @@ void PanelConfiguration::ShowRenderOptions()
...
+		ImGui::TextColored(ImVec4(1, 1, 1, 1), "and");
+		ImGui::SameLine();
+		ImGui::TextColored(ImVec4(0.3, 0.3, 0.3, 1), "Shadows");
+
+		ImGui::Separator();
+
+		ImGui::TextColored(ImVec4(1, 1, 0, 1), "Ambient Light");
+
+		ImGui::SliderFloat("Intensity", &App->lights->ambient_light_intensity, 0, 1, "%.2f");
+		ImGui::ColorEdit3("Color", App->lights->ambient_light_color);
```
Comment (mariofv) [inline Engine/EditorUI/Panel/PanelConfiguration.cpp:311]: This parameters of ambient light are not persistent, so we cannot change them and save the changes. To do that, we have to modify directly the class `ModuleLight`.

We should create a new light or find a way to store this values in  a file in order to modify them easily

File: Engine/Main/GameObject.cpp [hunk 18]
```diff
@@ -421,6 +421,9 @@ ENGINE_API Component* GameObject::CreateComponent(const Component::ComponentType
 	Component* created_component;
 	switch (type)
 	{
+	case Component::ComponentType::AABB:
+		created_component = new ComponentAABB(this);
+		break;
```
Comment (mariofv) [inline Engine/Main/GameObject.cpp:428]: This compoment is created automatically when creating `GameObject`, when are you using this case?

File: Engine/Module/ModuleDebugDraw.cpp [hunk 19]
```diff
@@ -396,7 +396,134 @@ bool ModuleDebugDraw::Init()
     APP_LOG_SUCCESS("Module Debug Draw initialized correctly.")
 
 	return true;
-}	
+}
+
+//void ModuleDebugDraw::Render()
```
Comment (marcorod94) [inline Engine/Module/ModuleDebugDraw.cpp]: Why is this huge amount of code commented?

File: Engine/Module/ModuleDebugDraw.cpp [hunk 20]
```diff
@@ -396,7 +396,134 @@ bool ModuleDebugDraw::Init()
...
+//	if (App->renderer->toggle_perspective_sub_frustums)
+//	{
+//		dd::frustum(App->cameras->camera_close->GetInverseClipMatrix(), float3(1, 1, 0));
+//		dd::frustum(App->cameras->camera_mid->GetInverseClipMatrix(), float3(0, 1, 1));
+//		dd::frustum(App->cameras->camera_far->GetInverseClipMatrix(), float3(1, 0, 1));
+//	}
+//
+//	RenderDebugDraws(*App->cameras->scene_camera);
+//
+//}
```
Comment (mariofv) [inline Engine/Module/ModuleDebugDraw.cpp]: Same as before, don't keep unnused comented code

File: Engine/Module/ModuleLight.cpp [hunk 21]
```diff
@@ -14,10 +17,37 @@ ModuleLight::~ModuleLight()
 bool ModuleLight::Init()
 {
 	APP_LOG_SECTION("************ Module Light Init ************");
+	directional_light_rotation = Quat::identity;
```
Comment (mariofv) [inline Engine/Module/ModuleLight.cpp]: Why this?

File: Engine/Module/ModuleLight.h [hunk 22]
```diff
@@ -3,6 +3,7 @@
 
 #include "Module.h"
 #include "Component/ComponentLight.h"
+#include "Component/ComponentAABB.h"
```
Comment (marcorod94) [inline Engine/Module/ModuleLight.h]: Code Style

File: Engine/Module/ModuleLight.cpp [hunk 23]
```diff
@@ -54,6 +86,9 @@ void ModuleLight::RenderDirectionalLight(const float3& mesh_position)
 
 		size_t light_direction_offset = App->program->uniform_buffer.lights_uniform_offset + 4 * sizeof(float);
 		glBufferSubData(GL_UNIFORM_BUFFER, light_direction_offset, sizeof(float3), light->owner->transform.GetFrontVector().ptr());
+	
+		directional_light_rotation = directional_light_rotation.LookAt(float3::unitZ, light->owner->transform.GetFrontVector(), float3::unitY, light->owner->transform.GetUpVector());
```
Comment (mariofv) [inline Engine/Module/ModuleLight.cpp:90]: Isn't this the same as `directional_light_rotation = light->owner->transform.GetFrontVector()` ?

File: Engine/Module/ModuleLight.cpp [hunk 24]
```diff
@@ -95,6 +130,49 @@ void ModuleLight::RenderSpotLights(const float3& mesh_position, GLuint program)
...
+		glUniformMatrix4fv(glGetUniformLocation(program, "main_cam_view"), 1, GL_TRUE, &App->cameras->main_camera->GetViewMatrix()[0][0]);
+		glUniformMatrix4fv(glGetUniformLocation(program, "main_cam_proj"), 1, GL_TRUE, &App->cameras->main_camera->GetProjectionMatrix()[0][0]);
+
+		glUniformMatrix4fv(glGetUniformLocation(program, "close_cam_view"), 1, GL_TRUE, &App->cameras->camera_close->GetViewMatrix()[0][0]);
+		glUniformMatrix4fv(glGetUniformLocation(program, "close_cam_proj"), 1, GL_TRUE, &App->cameras->camera_close->GetProjectionMatrix()[0][0]);
+
+		glUniformMatrix4fv(glGetUniformLocation(program, "mid_cam_view"), 1, GL_TRUE, &App->cameras->camera_mid->GetViewMatrix()[0][0]);
+		glUniformMatrix4fv(glGetUniformLocation(program, "mid_cam_proj"), 1, GL_TRUE, &App->cameras->camera_mid->GetProjectionMatrix()[0][0]);
+
+		glUniform1f(glGetUniformLocation(program, "main_cam_far_plane"), App->cameras->camera_close->camera_frustum.farPlaneDistance);
```
Comment (mariofv) [inline Engine/Module/ModuleLight.cpp]: Why are we passing 3 different cameras to render the main scene?

File: Engine/Module/ModuleWindow.cpp [hunk 25]
```diff
@@ -32,8 +32,8 @@ bool ModuleWindow::Init()
 		width = static_cast<int>(screen_width * 0.9f);
 		height = static_cast<int>(screen_height * 0.9f);
 		
-		//width = 1280;
-		//height = 720;
```
Comment (marcorod94) [inline Engine/Module/ModuleWindow.cpp:36]: remove comments

File: Engine/Module/ModuleRender.cpp [hunk 26]
```diff
@@ -158,6 +158,10 @@ void ModuleRender::Render() const
 	if (App->cameras->main_camera != nullptr) 
 	{
 		App->cameras->main_camera->RecordFrame(App->window->GetWidth(), App->window->GetHeight());
+		App->cameras->directional_light_camera->RecordFrame(App->window->GetWidth(), App->window->GetHeight());
+		App->cameras->directional_light_mid->RecordFrame(App->window->GetWidth(), App->window->GetHeight());
+		App->cameras->directional_light_far->RecordFrame(App->window->GetWidth(), App->window->GetHeight());
```
Comment (mariofv) [inline Engine/Module/ModuleRender.cpp]: In game the three different directional light frustums have the same resoultion, while in `PanelScene` and `PanelGame` don't. Why is that?

File: Engine/Module/ModuleRender.cpp [hunk 27]
```diff
@@ -195,6 +199,7 @@ void ModuleRender::RenderFrame(const ComponentCamera &camera)
 			mesh.second->Render();
 			num_rendered_tris += mesh.second->mesh_to_render->GetNumTriangles();
 			num_rendered_verts += mesh.second->mesh_to_render->GetNumVerts();
+			App->lights->UpdateLightAABB(*mesh.second->owner);
 			glUseProgram(0);
```
Comment (mariofv) [inline Engine/Module/ModuleRender.cpp:198]: While rendering directional light depth map the line `App->lights->UpdateLightAABB(*mesh.second->owner);` is called constantly, modyfing the light AABB when it shouldn't

File: Engine/Module/ModuleRender.h [hunk 28]
```diff
@@ -84,6 +84,10 @@ class ModuleRender : public Module
 
 public:
 	bool anti_aliasing = false;
+	bool toggle_ortho_frustum = false;
+	bool toggle_directional_light_aabb = true;
+	bool toggle_perspective_sub_frustums = false;
```
Comment (mariofv) [inline Engine/Module/ModuleRender.h:81]: This should be in `ModuleDebug`

File: Libraries/include/MathGeoLib/Geometry/Frustum.cpp [hunk 29]
```diff
@@ -203,7 +203,7 @@ float3 Frustum::FarPlanePos(float x, float y) const
 	else
 	{
 		float3 right = WorldRight();
-		return pos + front * farPlaneDistance
+		return pos + front * farPlaneDistance/2
```
Comment (mariofv) [inline Libraries/include/MathGeoLib/Geometry/Frustum.cpp]: Why this change?

**Review by mariofv — CHANGES_REQUESTED** [review]
Shadows look gorgeous! Few remarks:
- Performance is quite low, it went down from 60 fps to 45 fps. Looking at your code there are some inefficiencies that have to be adressed.
- One is just rendering the scene once, instead of having three diferent cameras. To solve what depth map to use when rendering shadows, use the depth of the fragment that is being rendered. (Look at Carlos slide, is specified there).
- Logic of creating depth maps is super mixed with the logic of rendering the scene normally, because `ComponentCamera` is being used for both of them. An example of this is calculating the directional light frustum when rendering the depth map (at this point directional light frustum should be already computed).
- There is a problem with shaders that is giving the following error:
```
Source: API Type: Undefined Behaviour Program undefined behavior warning: The current GL state uses a sampler (0) that has depth comparisons enabled, with a texture object (246) with a depth format, by a shader that samples it with a non-shadow sampler. This will result in undefined behavior
```
Because this is being logged in the Log all the time, the engine crashes after some time because the log runs out of memory
- Ambient light is too integrated in `ModuleLight`, and changes are not persistent

**Review by marcorod94 — CHANGES_REQUESTED** [review]
Please look at the changes in order to get a better performance

File: Engine/Module/ModuleDebugDraw.cpp (hunk 19 above)
Comment (tonifj) [reply Engine/Module/ModuleDebugDraw.cpp]: Some leftover code from the last merge

File: Engine/Module/ModuleLight.cpp (hunk 23 above)
Comment (tonifj) [reply Engine/Module/ModuleLight.cpp:90]: Probably (?)

File: Engine/Module/ModuleLight.cpp (hunk 24 above)
Comment (tonifj) [reply Engine/Module/ModuleLight.cpp]: There's no need to create 3 cameras because we only need the matrices with the cameras' traits. Now, we need a way to get those matrices without render the scene those many times.

File: Libraries/include/MathGeoLib/Geometry/Frustum.cpp (hunk 29 above)
Comment (tonifj) [reply Libraries/include/MathGeoLib/Geometry/Frustum.cpp]: Orthographic frustums, when rendered by dd::frustum, render its far plane at the double specified far plane distance, so, trying to fix it, I modified, this, but its already solved. So back to normal.

**Comment by mariofv** [comment]
Closing this because it was merged in #188
